### DISCUSSION TITLE
Base Repository for Symfony 6

### DISCUSSION THREAD

**Discussion by tavaresmatheus** [symfony/symfony#49563] [Q&A] [answered]
Hey there! I'm using the Base Repository architecture in my application, but I've reached a point that i'm stuck.
I have made a base repository that is working well in the application. But there is one method that isn't working and i have absolutely no idea on how to solve it. The method is update (update a generic entity on Doctrine) and i simply don't know how to do it generic, every example i've seen is to specific entities, not generic. This is the class i'm talking about:
```php
<?php

namespace App\Repository\Base;

use App\Repository\Base\Contracts\BaseRepositoryInterface;
use Doctrine\Bundle\DoctrineBundle\Repository\ServiceEntityRepository;
use Doctrine\Persistence\ManagerRegistry;

abstract class BaseRepository extends ServiceEntityRepository implements BaseRepositoryInterface
{
    protected string $entityClass;

    public function __construct(
        ManagerRegistry $registry,
        string $entityClass
    )
    {
        $this->entityClass = $entityClass;
        parent::__construct(
            $registry,
            $entityClass
        );
    }

    public function create(object $entity): array
    {
        $this->getEntityManager()->persist($entity);
        $this->getEntityManager()->flush();

        return [
            'id' => $entity->getId(),
        ];
    }

    public function findById(string $entityId): ?object
    {
        $entityFound = $this->getEntityManager()->find(
            $this->entityClass,
            $entityId
        );

        return $entityFound;
    }

    public function listAll(): array
    {
        $entities = $this->getEntityManager()->getRepository($this->entityClass);
        $entitiesList = $entities->findAll();

        return $entitiesList;
    }

    public function update(
        object $entity,
        array $dataToBeModified
    ): array
    {
        $queryBuilder = $this->getEntityManager()->createQueryBuilder();
        $result = $queryBuilder->update($this->entityClass, 'e')
            ->set($dataToBeModified, $dataToBeModified)
            ->where('e.id = ?')
            ->setParameter(0, $entity->getId())
            ->getQuery()
            ->execute();

        return [$result];
    }

    public function delete(string $entityId): bool
    {
        $entity = $this->getEntityManager()->find(
            $this->entityClass,
            $entityId
        );
        $this->getEntityManager()->remove($entity);
        $this->getEntityManager()->flush();

        return true;
    }
} 
```
If you have interest on see the repository, here it is: https://github.com/tavaresmatheus/localghost
Do you guys have any idea on how to do a update method to a generic entity?
Thank you very much in advance :)

**Comment by lyrixx** [comment]
> But there is one method that isn't working

What does it mean? What do you expect? What did you get? Have you a strack trace

OKay, found it
```
            ->set($dataToBeModified, $dataToBeModified)
```

---


It should be something like 
```
            ->set('property', 'value')
```

**Reply by tavaresmatheus** [reply]
Hi lyrix! yeah, you're right, but in the baserepository i don't know what property of the entity is going to me updated, so i'm really confused with that :/, i was trying to do in a generic way, but i have no idea by now.

**Comment by stof** [comment]
The implementation of your `listAll()` method looks suspicious to me. If your class extends `ServiceEntityRepository`, it means you plan to register it as the doctrine repository for your entity. So the code should use `$this->findAll()` instead of trying to get the repository again.

It looks like you wrote your code as if it was not the doctrine repository (which is also a valid approach, at least based on the provided info, but then it should not extend from the DoctrineBundle class)

**Reply by tavaresmatheus** [reply]
Hey there stof! you're correct about the `listAll()` method, i'm going to correct it. Thanks :).
The ideia behind it is to work as a Doctrine base repository, with all that repetitive CRUD stuff so i dont really need to do it every time i create an entity.

**Comment by lyrixx** [comment, marked as answer]
This thread is not directly related to your issue. But I strongly recommend you to **not** use this `BaseRepository`

1. This is useless
2. It's more code, so more bugs (you open an issues here, right 😄 )
3. It's missleading. For example, in the `create()` method, you call `flush()`. But it won't flush only the new entity, but every pending stash. This is how doctrine works.
4. You loose many tyehint. For example in `findById()` you return an object. However, with a more tradictionnal architecture you can return a `User` a `Project` etc. [example](https://github.com/symfony/demo/blob/main/src/Repository/PostRepository.php#L31)
5. There is unecessary code, like **all** the code `listAll()`. There is already a method `findAll()`  that does exactly the same thing;
6. There are extra query, like in delete(), you don't need the to fetch an object to delete it.

So I strongly advice you to remove this class, and use the default repository insteand. Your code will be simpler, with less code 🤗

**Reply by tavaresmatheus** [reply]
I'm new at Symfony, so maybe i'm doing things wrong HAHA. But i had some ideas. On Doctrine, when i create an entity there is always that `EntityRepository` by default, but in my mind i would have to write the same code to every entity i was creating, like the `CRUD` methods... so it would be the same CRUD for `Users`, then for `Events` and etc... i was trying just to write less code with the same benefit, but i guess i'm doing this wrong. There is another way? Or you think that just use the Doctrine Entity and replicate the code is the better way?
I really appreciate your help :)

**Reply by lyrixx** [reply]
I really think the default repository (generated by the maker bundle or not) is the way to go

You have everything you need in them and it's more standard.

You can have a look to the symfony/demo repo or the documentation

**Reply by tavaresmatheus** [reply]
Thanks, Lyrixx. I'm going that way!
